### ISSUE TITLE
Training the agent against 2 or more opponents + Max moves are stored and shouldn't

### ISSUE THREAD

**Issue by mancho1987** [hsahovic/poke-env#78]
Hey Haris,
I would like to train the agent against more than one opponent, but can't figure out how. 
Also, how could I make it train against itself and previous saved versions of it?

**Comment by hsahovic** [comment]
Hi @mancho1987,

Thanks for opening this issue.
I have a snippet somewhere that's doing precisely that - I'll try to find it tomorrow.

What is the precise behavior you'd like to setup?

Best,
Haris

**Comment by mancho1987** [comment]
Thanks for the quick reply @hsahovic!

I would like to try something like what they did for Alphastar, in which it trained against not just itself, but various opponents in a league. 
So for example, it would train first against random_player for 10000 steps, then against max_damage_player, and then against itself. This would be the initial setup, then I will build on it.

**Comment by hsahovic** [comment]
If you want to have several training sequences against different opponents, one simple approach would be to call your training function several times with these different opponents, ie.:

```python
for opponent in [random_opponent, max_dmg_opponent, shp_opponent]:
    env_player.play_against(
        env_algorithm=dqn_training,
        opponent=opponent,
        env_algorithm_kwargs={"dqn": dqn, "nb_steps": 100000},
    )
```

The next step would probably be including some custom logic in the way your training goes from one opponent to another.

If you want to get fancy, you can also create a custom training player that can alternate between different choice functions.
Let me know if that helps!

Best,
Haris

**Comment by mancho1987** [comment]
@hsahovic  Thanks helpful, thanks! 
In the case of the dqn example it works perfectly, but I tried to do this loop for the ppo workaround you suggested in another issue, and it didn't work. After training against the first opponent it just stops. How could this be fixed? 
I made a custom training player as well to compare the results. Will try it when the ppo multi training works :)
Btw, I want also to give a small reward for the amount of mons the current active kills, to try to encourage the discovery of setup and other sweeping strategies, but also am having trouble figuring out how to do that. Any pointers? 
Hope I'm not bothering too much! This is amazing and I am so excited to try a lot of things, unfortunately my programing skills are not great

**Comment by hsahovic** [comment]
Is any kind of arror logged after the training stops? If not, can you take a look at the battles going on locally (eg. on localhost) - is a battle with the second player starting and just hanging, or is it not even starting?

For the reward, you can define it however you want; the example implementation uses:

```python
    def compute_reward(self, battle) -> float:
        return self.reward_computing_helper(
            battle,
            fainted_value=2,
            hp_value=1,
            victory_value=30,
        )
```

`reward_computing_helper` exists to make creating custom rewards easier - i'd recommend taking a look at [its documentation](https://poke-env.readthedocs.io/en/stable/player.html?highlight=env%20player#poke_env.player.env_player.EnvPlayer.reward_computing_helper), as it provides a couple of options to customize your reward easily. In particular, you might be interested in playing with the `fainted_value` parameter.

If you want to do something more advanced, you might want to implement a fully custom reward; it that's the direction you want to take, the [actual implementation of `reward_computing_helper`](https://github.com/hsahovic/poke-env/blob/d18076a1ab93b4d26ee702f148ec63290766968f/src/poke_env/player/env_player.py#L180) might be a good starting point.

**Comment by mancho1987** [comment]
@hsahovic there is no error logged, it actually starts the first battle with the second opponent and just hangs there without making any moves. 
Got it about the rewards, thanks!

**Comment by hsahovic** [comment]
Hi @mancho1987,

Thanks for your patience. I uncovered an issue in `EnvPlayer`'s `reset` and `step` function, that are causing your issue. #82 should fix them shortly - I'll let you know when this is merged.

**Comment by hsahovic** [comment]
Version 0.3.8 should fix this :) Try running `pip install poke-env==0.3.8` and let me know how it goes!

**Comment by mancho1987** [comment]
Hi @hsahovic,
It works perfectly now, thanks! 
I wanted to ask a couple other questions:
- How do I get the base power of the opponent active moves? I am doing something like `for mon in opponent_team.values()
If mon.active:
For i, move in enumerate(mon.moves):
Opponent_moves_power[i] = move.base_power
` 
But I get an error like str has no attribute base_power. 
I would also like to get next all the moves from each of my mons and opponent mons, would it work the same way as for the active? 
- My feature vector is getting quite big, its taking my laptop several hours to train now. I am not experienced at all with training on aws or the likes. Is it possible to train it there? Could you give me a quick idea how I could do it?

**Comment by hsahovic** [comment]
Great, I'm glad to hear that!

- [`Pokemon.moves`](https://poke-env.readthedocs.io/en/stable/pokemon.html#poke_env.environment.pokemon.Pokemon.moves) is a `dict`: when you iterate over `mon.moves`, you iterate over the keys, not the `Move` objects; you need to replace that with `for i, move in enumerate(mon.moves.values())`. This should work the same for all mons.
- You can train models in the cloud, but I'm not well equipped to provide guidance on that subject. If I were you, I'd start by profiling my code to see if there are things that could be optimized beforehand, especially if you use a lot of native python data types.

**Comment by mancho1987** [comment]
Great! Thanks 
Btw, does mon.moves store all the moves a mon has used? Seems to me that it is storing max moves as well as regular moves. Is there a way that I can access only the available moves that every mon has, without temporary ones like max moves? 
Also, does mon stats include already the evs? I will try to make an agent for OU soon, and didn't find how to get the evs of each mon

**Comment by hsahovic** [comment]
Oh, it definitely shouldn't be storing max-moves: that's a bug. I'll try to get a fix for that.

Regarding evs: it's not storing raw evs, but it's storing your mon's raw stats (which take into account evs, ivs, level, nature etc.). In the future, I plan to add an estimation of your opponent's stats as well, based on damage calculations, but for now only base stats are available.

**Comment by hsahovic** [comment]
This should be fixed in version `0.4.1`.
